File: testsuites/HLT/Benchmark/concurrency-benchmark/EnterExit/go/EnterExit.go
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2026. All rights reserved.
 * This source file is part of the Cangjie project, licensed under Apache-2.0
 * with Runtime Library Exception.
 * 
 * See https://cangjie-lang.cn/pages/LICENSE for license information.
 */

package main

import (
  "fmt"
  "os"
  "runtime"
  "strconv"
  "sync"
  "time"
)

const FACTOR = 100

const (
  BIASED_LOCKING      = 0
  BACON_BITS          = 1
  MONITOR_UNCONTENDED = 2
  MONITOR_CONTENDED   = 3
)

var LockOSThread = false
var mode = BIASED_LOCKING
var nThreads = 1
var problemSize = 700000
var nSlicesForStat = 0

type Matrix struct {
  data    [][]int
  time    int
  maxTime int

  lastEnteredID    int
  ownerChanges     int
  timeSlice        int
  ownershipDistrib [][]int

  mu sync.Mutex
}

func newMatrix(nJobs int, jobLength int, nStatSlices int) *Matrix {
  data := make([][]int, nJobs)
  for i := range data {
    data[i] = make([]int, jobLength)
  }
  maxTime := nJobs * jobLength * FACTOR

  if nStatSlices != 0 {
    o := make([][]int, nStatSlices)
    for i := range o {
      o[i] = make([]int, nJobs)
    }
    return &Matrix{
      data:             data,
      maxTime:          maxTime,
      lastEnteredID:    -1,
      ownerChanges:     -1,
      ownershipDistrib: o,
      timeSlice:        maxTime / nStatSlices,
    }
  }

  return &Matrix{
    data:          data,
    maxTime:       maxTime,
    lastEnteredID: -1,
    ownerChanges:  -1,
  }
}

func (m *Matrix) postValidate() {
  if m.time != m.maxTime {
    fmt.Println("Validation failed: expected ", m.maxTime, " got ", m.time)
    os.Exit(1)
  }
}

func work(id int, m *Matrix, start, ready, finish *sync.WaitGroup) {
  if LockOSThread {
    runtime.LockOSThread()
  }
  ready.Done()
  start.Wait()

  for cnt := 0; cnt < FACTOR; cnt++ {
    len := len(m.data[id])
    for i := 0; i < len; i++ {
      m.mu.Lock()
      m.data[id][i] = id + i
      changed := (m.lastEnteredID != id)
      if changed {
        m.ownerChanges++
        m.lastEnteredID = id
      }
      if nSlicesForStat != 0 {
        slice := m.time / m.timeSlice
        m.ownershipDistrib[slice][id]++
      }
      m.time++
      m.mu.Unlock()
    }
  }

  finish.Done()
}

func bench(toPrint bool, sizeForAll int) {
  if mode == MONITOR_CONTENDED {
    if nThreads == 1 {
      fmt.Println("You opted for contended mode: thread number must be > 1")
      os.Exit(1)
    }
  } else {
    nThreads = 1
  }

  m := newMatrix(nThreads, sizeForAll/nThreads, nSlicesForStat)

  if mode == BACON_BITS {
    m.mu.Lock()
    m.mu.Unlock()
  } else if mode == MONITOR_UNCONTENDED {
    var latch sync.WaitGroup
    latch.Add(1)
    condition := sync.NewCond(&m.mu)
    go func() {
      condition.L.Lock()
      latch.Done()
      for true {
        // Under laboratory conditions of benchmark system this wait should not spuriously wake up.
        condition.Wait()
      }
      condition.L.Unlock()
    }()
    // Wait a bit to prevent interference of newly created thread.
    latch.Wait()
    time.Sleep(500 * time.Millisecond)
  }

  var start, ready, finish sync.WaitGroup
  start.Add(1)
  ready.Add(nThreads)
  finish.Add(nThreads)

  for i := 0; i < nThreads; i++ {
    go work(i, m, &start, &ready, &finish)
  }

  runtime.GC()

  ready.Wait()
  startTime := time.Now()
  start.Done()
  finish.Wait()
  time := time.Since(startTime).Milliseconds()

  m.postValidate()

  if toPrint {
    if nSlicesForStat != 0 {
      fmt.Println("Ownership distribution:")
      for slice := 0; slice < nSlicesForStat; slice++ {
        fmt.Print("|")
        for i := 0; i < nThreads; i++ {
          fmt.Print(" ", m.ownershipDistrib[slice][i], " |")
        }
        fmt.Println()
      }
    }
    fmt.Print("Ownership changes: ", m.ownerChanges)
    fmt.Print(" (GOMAXPROCS = ", runtime.GOMAXPROCS(0), ", LockOSThread = ", LockOSThread, ")")
    fmt.Println()

    fmt.Print("Time (workers = ", nThreads, " size = ", sizeForAll, "): ", time)
    fmt.Println()
  }
}

func parseArgs(args []string) {
  defer func() {
    if err := recover(); err != nil {
      fmt.Println("Usage: -biased | -bacon | -mon-uncontended | -mon-contended [-os-thread] [-size <value>] [-threads <num of threads>] [-scheduling-stat <num of slices>]")
      os.Exit(1)
    }
  }()

  for pos := 0; pos < len(args); pos++ {
    option := args[pos]
    if option == "-biased" {
      mode = BIASED_LOCKING
      continue
    } else if option == "-bacon" {
      mode = BACON_BITS
      continue
    } else if option == "-mon-uncontended" {
      mode = MONITOR_UNCONTENDED
      continue
    } else if option == "-mon-contended" {
      mode = MONITOR_CONTENDED
      continue
    } else if option == "-os-thread" {
      LockOSThread = true
      continue
    }

    pos++
    value, err := strconv.Atoi(args[pos])

    if value <= 0 || err != nil {
      panic("value invalid")
    }

    if option == "-size" {
      problemSize = value
    } else if option == "-threads" {
      nThreads = value
    } else if option == "-scheduling-stat" {
      nSlicesForStat = value
    } else {
      panic("invalid option")
    }
  }
}

func main() {

  args := os.Args
  parseArgs(args[1:])

  bench(false, 5000)
  bench(true, problemSize)
}

File: testsuites/HLT/Benchmark/concurrency-benchmark/EnterExit/loom/EnterExit.java
/*
 * Copyright (c) Huawei Technologies Co., Ltd. 2026. All rights reserved.
 * This source file is part of the Cangjie project, licensed under Apache-2.0
 * with Runtime Library Exception.
 * 
 * See https://cangjie-lang.cn/pages/LICENSE for license information.
 */

import java.util.concurrent.CountDownLatch;

public class EnterExit {
    static final int FACTOR = 100;  //to multiply problem size

    static final class Matrix {
        int[][] data;
        int time = 0;
        int maxTime;

        // ownership stat
        int lastEnteredId = -1;  // invalid
        int ownerChanges = -1;
        int timeSlice;
        int[][] ownershipDistrib;

        ////////////////

        Matrix(int nJobs, int jobLength, int nStatSlices) {
            this.data = new int[nJobs][jobLength];
            this.maxTime = nJobs * jobLength * FACTOR;

            if (nStatSlices != 0) {
                this.ownershipDistrib = new int[nStatSlices][nJobs];
                this.timeSlice = this.maxTime / nStatSlices;
            }
        }

        void postValidate() {
            if (this.time != this.maxTime) {
                System.out.println("Validation failed: expected " + this.maxTime + " got " + this.time);
                System.exit(1);
            }
        }
    }

    static class Worker implements Runnable {
        int id;
        final Matrix m;
        CountDownLatch start, ready, finish;

        Worker(int id, Matrix m, CountDownLatch start, CountDownLatch ready, CountDownLatch finish) {
            this.id = id;
            this.m = m;
            this.start = start;
            this.ready = ready;
            this.finish = finish;
        }

        public void run() {
            try {
                ready.countDown();
                start.await();

                for (int cnt = 0; cnt < FACTOR; cnt++) {
                    int len = m.data[id].length;
                    for (int i = 0; i < len; i++) {
                 
                        synchronized (m) {
                            m.data[id][i] = id + i;

                            boolean changed = (m.lastEnteredId != id);
                            if (changed) {
                                m.ownerChanges++;
                                m.lastEnteredId = id;
                            }
                            if (nSlicesForStat != 0) {
                                int slice = m.time / m.timeSlice;
                                m.ownershipDistrib[slice][id]++;
                            }
                            m.time++;
                        }
                    }
                }

                finish.countDown();
                finish.await();
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    enum Mode {BIASED_LOCKING, BACON_BITS, MONITOR_UNCONTENDED, MONITOR_CONTENDED}

    static Mode mode           = Mode.BIASED_LOCKING;
    static int  nThreads       = 1;
    static int  problemSize    = 700000;
    static int  nSlicesForStat = 0;

    static final volatile Matrix hide;
    static volatile Thread t;

    static void prepareMatrix(Matrix m) {
        if (mode == Mode.BACON_BITS) {
            synchronized (m) {};                        // false biasing
        } else if (mode == Mode.MONITOR_UNCONTENDED) {
            try {
                CountDownLatch latch = new CountDownLatch(1);
                // deflation prevention measure
                t = new Thread(() -> {
                    try {
			                  synchronized (hide) {
                    	      latch.countDown();
                            while (true) {
                                // Under laboratory conditions of benchmark system this wait should not spuriously wake up.
                                m.wait();
                            }
			                  }
                    } catch (InterruptedException e) {}
                });
            		t.setDaemon(true);
            		t.start();
                latch.await();
                // Wait a bit to prevent interference of newly created thread.
                Thread.sleep(500);
            } catch (InterruptedException e) {}
        }
    }

    static void bench(boolean toPrint, int sizeForAll) {
        if (mode == Mode.MONITOR_CONTENDED) {
            if (nThreads == 1) {
                System.out.println("You opted for contended mode: thread number must be > 1");
                System.exit(1);
            }
        } else
            nThreads = 1;

        // intentionally assign into static volatile variable to confuse partial escape analysis and prevent optimizing out some ifs inside `prepareMatrix` 
        hide = new Matrix(nThreads, sizeForAll / nThreads, nSlicesForStat);
        prepareMatrix(hide);

        Matrix m = hide;

        CountDownLatch start = new CountDownLatch(1);
        CountDownLatch ready = new CountDownLatch(nThreads);
        CountDownLatch finish= new CountDownLatch(nThreads);

        for (int i = 0; i < nThreads; i++) {
            Thread.ofVirtual().start(new Worker(i, m, start, ready, finish));
        }

        System.gc();

        try {
            ready.await();
            long startTime = System.currentTimeMillis();
            start.countDown();
            finish.await();

            long time = System.currentTimeMillis() - startTime;

            m.postValidate();

            if (toPrint) {
                if (nSlicesForStat != 0) {
                    System.out.println("Ownership distribution:");
                    for (int slice = 0; slice < nSlicesForStat; slice++) {
                        System.out.print("|");
                        for (int i = 0; i < nThreads; i++) {
                            System.out.print(" " + m.ownershipDistrib[slice][i] + " |");
                        }
                        System.out.println("");
                    }
                }
                System.out.println("Ownership changes: " + m.ownerChanges);
                System.out.println("Time (workers = " + nThreads + " size = " + sizeForAll + "): " + time);
            }
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    static void parseArgs(String args[]) throws Exception {
        for (int pos = 0; pos < args.length; pos++) {
            String option = args[pos];
            if (option.equals("-biased")) {
                mode = Mode.BIASED_LOCKING;
                continue;
            } else if (option.equals("-bacon")) {
                mode = Mode.BACON_BITS;
                continue;
            } else if (option.equals("-mon-uncontended")) {
                mode = Mode.MONITOR_UNCONTENDED;
                continue;
            } else if (option.equals("-mon-contended")) {
                mode = Mode.MONITOR_CONTENDED;
                continue;
            }

            pos++;
            int value = Integer.parseInt(args[pos]);

            if (value <= 0)
                throw new Exception();

            if (option.equals("-size")) {
                problemSize = value;
            } else if (option.equals("-threads")) {
                nThreads = value;
            }else if (option.equals("-scheduling-stat")) {
                nSlicesForStat = value;
            } else
                throw new Exception();
        }
    }

    public static void main(String[] args) {
        try {
            parseArgs(args);
        } catch (Exception e) {
            System.out.println("Usage: -biased | -bacon | -mon-uncontended | -mon-contended [-size <value>] [-threads <num of threads>] [-scheduling-stat <num of slices>]");
            System.exit(1);
        }

        bench(false, 5000);        //warm-up for "those JVM that need it"
        bench(true, problemSize);
    }
}
